crates/accelerate/src/target_transpiler/mod.rs

```diff
@@ -1307,3 +1376,118 @@ pub fn target(m: &Bound<PyModule>) -> PyResult<()> {
 }
 
 // TODO: Add rust-based unit testing.
+#[cfg(all(test, not(miri)))]
+mod test {
+    use std::sync::Mutex;
+
+    use crate::target_transpiler::{Target, TargetOperation};
+    use nalgebra::Matrix2;
+    use numpy::Complex64;
+    use pyo3::{exceptions::PyRuntimeError, prelude::*};
+    use qiskit_circuit::{
+        imports::ImportOnceCell,
+        operations::{Param, StandardGate, StandardInstruction, UnitaryGate},
+    };
+    use qiskit_circuit::{imports::UNITARY_GATE, packed_instruction::PackedOperation};
+    use smallvec::smallvec;
+
+    static PY_RXGATE: ImportOnceCell = ImportOnceCell::new("qiskit.circuit.library", "RXGate");
+    static BARRIER: ImportOnceCell = ImportOnceCell::new("qiskit.circuit.library", "Barrier");
+    static IMPORT_LOCK: Mutex<()> = Mutex::new(());
+
+    #[test]
+    fn add_standard_gate_without_python() -> PyResult<()> {
+        let _lock = IMPORT_LOCK
```

Comment:
There was some discussion a while back about having `qiskit` get imported up front to avoid this sort of thing:

https://github.com/Qiskit/qiskit/pull/13776#issuecomment-2637618226

I didn't bother submitting a PR to do that since there's also the issue of `PyObject`s not being extractable to Rust `pyclass` structs due to ABI compat issues (i.e. in the Rust tests, our `pyclass` structs come from the `rlib`, but the structs that back `PyObject`s coming from the actual Qiskit code originate from the `cdylib` linked into the Python process).

My assumption was that folks probably would just not really use PyO3 Rust unit testing for anything complex. But, it looks like you have a use case here. Should I make a PR to avoid the deadlock issue?

Reply:
If it does solve the issue, I would love to have a PR that does that. It would be pretty helpful.

Reply:
Done in #14200.

I tried it with this PR removing the mutex and it seemed to work 🙂/

Reply:
I commented on #14200, but just to centralise: if we do this, I'd agree that #14200 is a better strategy. That said, I'm not convinced that we ought to use Python in this test. I'm not entirely sure what the use of Python is buying us - if we test the Rust objects from Rust, is there a really serious risk that "object created from Rust" might not produce the correct Python object later in a way that would not be caught by regular tests of the creation of Python-free `StandardGate` objects?

Even if this test can in theory work at this moment in time, I'm not crazy about doing more testing where `_accelerate` is imported from apython in Rust, because we know that that's not ABI compatible and feels like an accident waiting to happen.

Reply:
IMO, #14200 is a nice-to-have if we're going to keep PyO3 unit testing around for the time being. The main use-case I'd see for it is as a stop-gap measure for cases where our internal Rust code relies on importing parts of our data model that aren't yet native to Rust.

But in this particular case, I'd agree that we're not getting much by testing that the Rust-backed operations from the target are equivalent to gates constructed via Python. I'd rather expect any testing here to be poking at the `Target` Rust instance to make sure the instructions were added successfully (no Python involved), if you're to have any testing here at all.